neurom/apps/cut_plane_detection.py

```diff
@@ -0,0 +1,207 @@
+#        notice, this list of conditions and the following disclaimer.
+#     2. Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in the
+#        documentation and/or other materials provided with the distribution.
+#     3. Neither the name of the copyright holder nor the names of
+#        its contributors may be used to endorse or promote products
+#        derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+'''Module for the detection of the cut plane'''
+import logging
+import operator
+
+import numpy as np
+import neurom as nm
+
+from neurom import viewer
+from neurom.core import Tree
+from neurom.core.dataformat import COLS
+
+
+L = logging.getLogger(__name__)
+
+
+def _create_1d_distributions(points, bin_width):
+    '''Create point histograms along each axis
+
+    Parameters:
+        points: a np.ndarray of points
+        bin_width: the bin width
+
+    Returns: a dict of the X, Y and Z 1D histograms'''
+    min_, max_ = np.min(points, axis=0), np.max(points, axis=0)
+    hist = dict()
+    for i, plane in enumerate('XYZ'):
+        # Two binnings with same spacing but different offsets
+        # first is attached to min_[i], last is attached to max_[i]
+        binning_first_slice = np.arange(min_[i], max_[i] + bin_width, bin_width)
+        binning_last_slice = np.arange(max_[i], min_[i] - bin_width, -bin_width)[::-1]
+        hist[plane] = (np.histogram(points[:, i], bins=binning_first_slice),
+                       np.histogram(points[:, i], bins=binning_last_slice))
+    return hist
+
+
+def _get_probabilities(hist):
+    '''Returns -log(p) where p is the a posteriori probabilities of the observed values
+    in the bins X min, X max, Y min, Y max, Z min, Z max
+
+    Parameters:
+        hist: a dict of the X, Y and Z 1D histograms
+
+    Returns: a dict of -log(p) values'''
+    def minus_log_p(mu):
+        '''Compute -Log(p) where p is the a posteriori probability to observe 0 counts
+        in bin given than the mean value was "mu":
+        The number of counts follows a Poisson law so the result is simply... mu
```

Comment:
the counts is assumed to follow Poisson's law, right?

Reply:
Yep, I think it does. Don't you think ?

Reply:
Dunno, haven't thought about it enough